core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -333,6 +346,181 @@ public long nextFetchOffset() {
         }
     }
 
+    private void updateOffsetsOnLsoMovement() {
+        long logStartOffset = offsetForEarliestTimestamp();
```

Comment:
Is this a heavy burden to call this on every fetch?

Reply:
Completely agree, I think we should fetch this only when exception occurs.